### PR TITLE
Vinu/DTRA-1444/feat: added generic wheel picker component

### PR DESCRIPTION
**Pull request by vinu-deriv**
Added generic wheel picker

Screenshot 2024-07-30 at 1 38 33 PM

### CODE REVIEW

File: lib/components/Atom/WheelPicker/index.tsx [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    selectedValue: string | number;
+    setSelectedValue: (value: string | number) => void;
+    handleKeyDown: (e: React.KeyboardEvent) => void;
+    focus: boolean;
+    dropDownItemClassName?: string;
+    containerClassName?: string;
+    listClassName?: string;
+}
+
+export const WheelPicker = ({
```
Comment (maryia-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:25]: @vinu-deriv let's have 1 component per file :)🙏🏻

**Review by maryia-deriv — CHANGES_REQUESTED** [review]

File: lib/components/Atom/WheelPicker/index.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            ),
+        [data],
+    );
+
+    const currentDataValue = useRef(dataItemsMap.get(selectedValue) ?? 0);
+    const dataItemsContRef = useRef<HTMLUListElement>(null);
+
+    const handleDataScroll = (event: Event) => {
+        let isAnimating = false;
+        if (!isAnimating) {
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:51]: Sorry, didn't get: we are assign a value (`false`) to a variable (`isAnimating`) and then checking it?

Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx:51]: I will remove this , had forgotten to remove it . I was trying to put this code as callback for requestAnimationFrame, so had added it

File: lib/components/Atom/WheelPicker/index.tsx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    useEffect(() => {
+        if (!dataItemsContRef.current) return;
+        setSelectedValue(data[currentDataValue.current].value);
+        const divHeight = dataItemsContRef?.current.clientHeight;
+        resizeObserver.observe(dataItemsContRef.current);
+        dataItemsContRef.current.style.paddingTop = `${divHeight * 0.5}px`;
+        dataItemsContRef.current.style.paddingBottom = `${divHeight * 0.5}px`;
+        dataItemsContRef.current.addEventListener("scroll", handleDataScroll);
+        return () => {
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:91]: I think we need to `unobserve` or `disconnect` from `dataItemsContRef.current` here too

File: lib/components/Atom/WheelPicker/index.tsx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+import clsx from "clsx";
+import "./wheel-picker.scss";
+import { InputProps } from "@components/Input/base";
+
+export interface WheelPickerProps extends HTMLAttributes<HTMLElement> {
+    data: { value: string | number }[];
+    selectedValue: string | number;
+    setSelectedValue: (value: string | number) => void;
+    handleKeyDown: (e: React.KeyboardEvent) => void;
+    focus: boolean;
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:19]: I'm not the right person, who can give advise regarding naming, but maybe for boolean it'll be better to use `isFocused`? Or something that sound more boolean-like?

File: lib/components/Atom/WheelPicker/index.tsx [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                                dropDownItemClassName,
+                            )}
+                            label={value}
+                            role="option"
+                            tabIndex={0}
+                            textAlignment="center"
+                            ref={(node: HTMLDivElement) =>
+                                (dataRefs.current[index] = node)
+                            }
+                            key={index}
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:135]: Is there any error regarding index as keys? Maybe we can use value instead? Please, ignore if irrelevant

File: lib/components/Atom/WheelPicker/index.tsx [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            });
+        }
+    });
+
+    useEffect(() => {
+        if (!dataItemsContRef.current) return;
+        setSelectedValue(data[currentDataValue.current].value);
+        const divHeight = dataItemsContRef?.current.clientHeight;
+        resizeObserver.observe(dataItemsContRef.current);
+        dataItemsContRef.current.style.paddingTop = `${divHeight * 0.5}px`;
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:88]: It looks a bit similar to what is happening on lines 73-75. Please, ignore if irrelevant

File: lib/components/Atom/WheelPicker/index.tsx [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        if (setSelectedValue)
+            setSelectedValue(
+                `${dataValues[0] ?? ""} ${dataValues[1] ?? ""} ${dataValues[2] ?? ""}`,
+            );
+    }, [dataValues]);
+
+    const handleKeyDown = (e: KeyboardEvent<Element>, index: number) => {
+        setColRef(new Array(data.length).fill(false));
+        switch (e.key) {
+            case "ArrowRight":
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:186]: In _lib/utils/common-utils.ts_ there is an object for key names) You can reuse it if you want)

File: lib/components/Atom/WheelPicker/wheel-picker.scss [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    &__data-items {
+        width: 100%;
+        height: 100%;
+        margin: 0;
+        overflow-y: scroll;
+        scroll-snap-type: y mandatory;
+        padding: 0;
+        -ms-overflow-style: none;
+        scrollbar-width: none;
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/wheel-picker.scss:35]: Please, check if will work in Real IOS (Safari). In my previous PR I found out, that `scrollbar-width` css property is not working there, so I added 
```
            &::-webkit-scrollbar {
                display: none; /* Safari */
            }
![Uploading Screenshot 2024-07-30 at 2.44.16 PM.png…]()

```

Comment (kate-deriv) [reply lib/components/Atom/WheelPicker/wheel-picker.scss:35]: Screenshot 2024-07-30 at 2 44 16 PM

File: lib/components/Atom/WheelPicker/wheel-picker.scss [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        transition: all var(--core-motion-duration-200)
+            var(--core-motion-ease-400);
+        &::before {
+            position: absolute;
+            content: "";
+            transform: translateY(-50%);
+            width: 100%;
+            height: 48px;
```
Comment (kate-deriv) [inline lib/components/Atom/WheelPicker/wheel-picker.scss:43]: --core-spacing-2400 or --core-size-2400 will also look nice here

File: lib/components/Atom/WheelPicker/wheel-picker.scss (hunk 8 above)
Comment (kate-deriv) [reply lib/components/Atom/WheelPicker/wheel-picker.scss:35]: Oh, sorry, found it below, ignore my comment)))))

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+        const handleKeyDownEvent = (e: KeyboardEvent<HTMLDivElement>) => {
+            if (e.key === "ArrowDown") {
+                open();
+            }
+            
+        };
+
+        useEffect(() => {
+            if (onValueChange) onValueChange(daataValue);
```
Comment (kate-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:83]: daataValue --> dataValue?

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                                        {...rest}
+                                    >
+                                        {children}
+                                    </WheelPickerContainer>
+                                </div>
+                            )
+                        ) : (
+                            <ActionSheet.Root
+                                isOpen={isOpen}
+                                onClose={() => close()}
```
Comment (kate-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:136]: Will it work like this: onClose={close} ? Please, ignore if irrelevant

File: lib/components/Atom/WheelPicker/index.tsx [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+    );
+
+    const currentDataValue = useRef(dataItemsMap.get(selectedValue) ?? 0);
+    const dataItemsContRef = useRef<HTMLUListElement>(null);
+
+    const handleDataScroll = (event: Event) => {
+        let isAnimating = false;
+        if (!isAnimating) {
+            isAnimating = true;
+            if (!event.target) return;
```
Comment (henry-deriv) [inline lib/components/Atom/WheelPicker/index.tsx:53]: should this be moved up to line 50?

File: lib/components/Atom/WheelPicker/base/wheel-picker.scss [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    &__data-items {
+        width: 100%;
+        height: 100%;
+        margin: 0;
+        overflow-y: scroll;
+        scroll-snap-type: y mandatory;
+        padding: 0;
+        -ms-overflow-style: none;
+        scrollbar-width: none;
+        overflow-wrap:anywhere ;
```
Comment (henry-deriv) [inline lib/components/Atom/WheelPicker/base/wheel-picker.scss:39]:
```suggestion
        overflow-wrap: anywhere;
```

File: lib/components/Atom/WheelPicker/index.tsx (hunk 12 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx]: yes made the change

File: lib/components/Input/generic-wheel-picker/index.tsx (hunk 11 above)
Comment (vinu-deriv) [reply lib/components/Input/generic-wheel-picker/index.tsx]: yes made the change

File: lib/components/Input/generic-wheel-picker/index.tsx (hunk 10 above)
Comment (vinu-deriv) [reply lib/components/Input/generic-wheel-picker/index.tsx]: thanks for noticing it

File: lib/components/Atom/WheelPicker/index.tsx (hunk 6 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx]: yes thanks for noticing, made the clean up had forgotten to remove

File: lib/components/Atom/WheelPicker/index.tsx (hunk 5 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx]: yes it has to be value

File: lib/components/Atom/WheelPicker/index.tsx (hunk 4 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx]: yes made the change

File: lib/components/Atom/WheelPicker/index.tsx (hunk 3 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/index.tsx]: have removed the listeners and disconnected from the observer inside useEffect return

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        };
+
+        const handleKeyDownEvent = (e: KeyboardEvent<HTMLDivElement>) => {
+            if (e.key === KEY.ARROW_DOWN) {
+                open();
+            }
+        };
+
+        useEffect(() => {
+            if (onValueChange) onValueChange(inputValues);
```
Comment (kate-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:100]: onValueChange?.(inputValues);

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+        const handleMouseDown = () => {
+            setIsPressed(true);
+        };
+
+        const handleMouseUp = () => {
+            setIsPressed(false);
+        };
+
+        const handleMouseLeave = () => {
```
Comment (kate-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:64]: handleMouseUp  and handleMouseLeave do the same and all of 3 functions can be passed to the props as () => setIsPressed(...)

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+                        onMouseUp={handleMouseUp}
+                        onMouseLeave={handleMouseLeave}
+                    >
+                        <Input
+                            dropdown
+                            isDropdownOpen={isOpen}
+                            value={selectedValue ?? initialValues}
+                            autoFocus={false}
+                            autoComplete="off"
+                            readOnly={true}
```
Comment (kate-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:126]: can omit `true`

File: lib/components/Input/generic-wheel-picker/index.tsx (hunk 14 above)
Comment (kate-deriv) [reply lib/components/Input/generic-wheel-picker/index.tsx:100]: Absolutely optional)

File: lib/components/Input/generic-wheel-picker/index.tsx (hunk 16 above)
Comment (kate-deriv) [reply lib/components/Input/generic-wheel-picker/index.tsx:126]: Absolutely optional)

File: lib/components/Input/generic-wheel-picker/index.tsx [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+                                    </WheelPickerContainer>
+                                </div>
+                            )
+                        ) : (
+                            <ActionSheet.Root isOpen={isOpen} onClose={close}>
+                                <ActionSheet.Portal
+                                    shouldCloseOnDrag={true}
+                                    fullHeightOnOpen={true}
+                                    ref={actionSheetRef}
+                                >
```
Comment (prince-deriv) [inline lib/components/Input/generic-wheel-picker/index.tsx:163]: @jia-deriv looks like `isMobile` hook achieves thesame purpose of the `breakpointProvider` can you confirm?

**Review by maryia-deriv — APPROVED** [review]

**Review by ahmadtaimoor-deriv — APPROVED** [review]

File: lib/components/Atom/WheelPicker/base/index.tsx [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+export interface WheelPickerProps extends HTMLAttributes<HTMLElement> {
+    data: { value: string | number }[];
+    selectedValue: string | number;
+    setSelectedValue: (value: string | number) => void;
+    handleKeyDown?: (e: React.KeyboardEvent) => void;
+    isFocused?: boolean;
+    dropDownItemClassName?: string;
+    containerClassName?: string;
+    listClassName?: string;
+    position?: "first" | "center" | "last" | undefined;
```
Comment (prince-deriv) [inline lib/components/Atom/WheelPicker/base/index.tsx:15]: Can we make use of `TPosition` here?

File: lib/components/Atom/WheelPicker/base/index.tsx [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        });
+        setSelectedValue(data[selectedElement].value);
+    };
+
+    const resizeObserver = new ResizeObserver(() => {
+        const index = Number(currentDataValue.current);
+        if (!itemsContainerRef?.current) return;
+        const divHeight = itemsContainerRef?.current.clientHeight;
+        itemsContainerRef.current.style.paddingTop = `${divHeight * 0.5}px`;
+        itemsContainerRef.current.style.paddingBottom = `${divHeight * 0.5}px`;
```
Comment (prince-deriv) [inline lib/components/Atom/WheelPicker/base/index.tsx:64]: Can `0.5` be a variable multiplier so we can easily update later incase of UI change.

File: lib/components/Atom/WheelPicker/base/wheel-picker.scss [hunk 20]
```diff
@@ -0,0 +1,87 @@
+@import "@deriv-com/quill-tokens/dist/breakpoints.scss";
+
+.quill-wheel-picker {
+    display: flex;
+    flex-direction: row;
+    align-items: center;
+    justify-content: space-evenly;
+
```
Comment (prince-deriv) [inline lib/components/Atom/WheelPicker/base/wheel-picker.scss:8]:
```suggestion
```

File: lib/components/Atom/WheelPicker/base/wheel-picker.scss [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+    height: 100%;
+    background-color: var(--component-field-bg-outline-default);
+    padding: 0;
+    @include breakpoint("md") {
```
Comment (prince-deriv) [inline lib/components/Atom/WheelPicker/base/wheel-picker.scss:12]:
```suggestion
    padding: 0;
    
    @include breakpoint("md") {
```

File: lib/components/Atom/WheelPicker/base/index.tsx (hunk 18 above)
Comment (vinu-deriv) [reply lib/components/Atom/WheelPicker/base/index.tsx]: I will use THorizontalPosition, that is a better suit

**Review by prince-deriv — APPROVED** [review]

**Pull request merged**